fuse/opcode.go

```diff
@@ -442,7 +443,7 @@ func doStatFs(server *Server, req *request) {
 }
 
 func doIoctl(server *Server, req *request) {
-	req.status = ENOSYS
+	req.status = Status(syscall.ENOTTY)
```

Comment:
Returning ENOSYS makes the kernel not send any further IOCTL requests to us, right?

ENOTTY does not do that, I guess. Have seen any consequences?

Reply:
@rfjakob Thank you for taking a look at this patch.

> Returning ENOSYS makes the kernel not send any further IOCTL requests to us, right?

Kernel seems to send IOCTL repeatedly even using ENOSYS, same as ENOTTY.
 So, at least from this perspective, ENOTTY doesn't seem to have downsides against ENOSYS.


Repro:
When mounting [hello filesystem](https://github.com/hanwen/go-fuse/blob/master/example/hello/main.go) to `/mnt/hello` and running `lsattr /mnt/hello ; lsattr /mnt/hello ; lsattr /mnt/hello`, IOCTL is invoked repeatedly.

- Ubuntu 20.04 + kernel v5.16-rc7

```
# go run ./example/hello/main.go -debug /mnt/hello
2021/12/28 01:48:00 callFusermount: executing ["/bin/fusermount" "/mnt/hello" "-o" "subtype=rawBridge"]
2021/12/28 01:48:00 rx 2: INIT n0 {7.35 Ra 0x20000 ASYNC_READ,BIG_WRITES,SPLICE_MOVE,READDIRPLUS_AUTO,EXPLICIT_INVAL_DATA,NO_OPEN_SUPPORT,POSIX_ACL,ATOMIC_O_TRUNC,DONT_MASK,SPLICE_WRITE,SPLICE_READ,AUTO_INVAL_DATA,NO_OPENDIR_SUPPORT,EXPORT_SUPPORT,FLOCK_LOCKS,READDIRPLUS,WRITEBACK_CACHE,PARALLEL_DIROPS,CACHE_SYMLINKS,POSIX_LOCKS,IOCTL_DIR,ASYNC_DIO,HANDLE_KILLPRIV,ABORT_ERROR,MAX_PAGES,0x30000000} 
2021/12/28 01:48:00 tx 2:     OK, {7.28 Ra 0x20000 BIG_WRITES,ASYNC_READ,AUTO_INVAL_DATA,NO_OPEN_SUPPORT,READDIRPLUS,PARALLEL_DIROPS 0/0 Wr 0x10000 Tg 0x0}
2021/12/28 01:48:00 rx 4: LOOKUP n1 [".go-fuse-epoll-hack"] 20b
2021/12/28 01:48:00 tx 4:     OK, {n18446744073709551615 g0 tE=0s tA=0s {M0100644 SZ=0 L=1 0:0 B0*0 i0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:00 rx 6: OPEN n18446744073709551615 {O_RDONLY,0x8000} 
2021/12/28 01:48:00 tx 6:     OK, {Fh 18446744073709551615 }
2021/12/28 01:48:00 rx 8: POLL n18446744073709551615 
2021/12/28 01:48:00 tx 8:     38=function not implemented
2021/12/28 01:48:00 rx 10: FLUSH n18446744073709551615 {Fh 18446744073709551615} 
2021/12/28 01:48:00 tx 10:     OK
2021/12/28 01:48:00 rx 12: RELEASE n18446744073709551615 {Fh 18446744073709551615 0x8000  L0} 
2021/12/28 01:48:00 tx 12:     OK
2021/12/28 01:48:28 rx 14: GETATTR n1 {Fh 0} 
2021/12/28 01:48:28 tx 14:     OK, {tA=0s {M040755 SZ=0 L=0 0:0 B0*4096 i0:0 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 16: STATFS n1 
2021/12/28 01:48:28 tx 16:     OK, {blocks (0,0)/0 files 0/0 bs0 nl0 frs0}
2021/12/28 01:48:28 rx 18: OPENDIR n1 
2021/12/28 01:48:28 tx 18:     OK, {Fh 1 }
2021/12/28 01:48:28 rx 20: GETATTR n1 {Fh 0} 
2021/12/28 01:48:28 tx 20:     OK, {tA=0s {M040755 SZ=0 L=0 0:0 B0*4096 i0:0 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 22: READDIRPLUS n1 {Fh 1 [0 +4096)  L 0 NONBLOCK,DIRECTORY,0x8000} 
2021/12/28 01:48:28 tx 22:     OK,  160b data "\x02\x00\x00\x00\x00\x00\x00\x00"...
2021/12/28 01:48:28 rx 24: LOOKUP n1 ["file.txt"] 9b
2021/12/28 01:48:28 tx 24:     OK, {n2 g0 tE=0s tA=0s {M0100644 SZ=8 L=0 0:0 B8*4096 i0:2 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 26: GETATTR n2 {Fh 0} 
2021/12/28 01:48:28 tx 26:     OK, {tA=0s {M0100644 SZ=8 L=0 0:0 B8*4096 i0:2 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 28: LOOKUP n1 ["file.txt"] 9b
2021/12/28 01:48:28 tx 28:     OK, {n2 g0 tE=0s tA=0s {M0100644 SZ=8 L=0 0:0 B8*4096 i0:2 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 30: OPEN n2 {NONBLOCK,0x28000} 
2021/12/28 01:48:28 tx 30:     OK, {Fh 0 CACHE}
2021/12/28 01:48:28 rx 32: OPEN n2 {O_RDONLY} 
2021/12/28 01:48:28 tx 32:     OK, {Fh 0 CACHE}
2021/12/28 01:48:28 rx 34: IOCTL n2 *fuse._IoctlIn: &{{72 39 34 2 {{0 0} 836} 0} 0 0 2148034049 0 0 4} 
2021/12/28 01:48:28 tx 34:     38=function not implemented
2021/12/28 01:48:28 rx 36: RELEASE n2 {Fh 0   L0} 
2021/12/28 01:48:28 tx 36:     OK
2021/12/28 01:48:28 rx 38: FLUSH n2 {Fh 0} 
2021/12/28 01:48:28 tx 38:     OK
2021/12/28 01:48:28 rx 40: RELEASE n2 {Fh 0 NONBLOCK,0x28000  L0} 
2021/12/28 01:48:28 tx 40:     OK
2021/12/28 01:48:28 rx 42: READDIRPLUS n1 {Fh 1 [1 +4096)  L 0 NONBLOCK,DIRECTORY,0x8000} 
2021/12/28 01:48:28 tx 42:     OK
2021/12/28 01:48:28 rx 44: RELEASEDIR n1 {Fh 1 NONBLOCK,DIRECTORY,0x8000  L0} 
2021/12/28 01:48:28 tx 44:     OK
2021/12/28 01:48:28 rx 46: GETATTR n1 {Fh 0} 
2021/12/28 01:48:28 tx 46:     OK, {tA=0s {M040755 SZ=0 L=0 0:0 B0*4096 i0:0 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 48: STATFS n1 
2021/12/28 01:48:28 tx 48:     OK, {blocks (0,0)/0 files 0/0 bs0 nl0 frs0}
2021/12/28 01:48:28 rx 50: OPENDIR n1 
2021/12/28 01:48:28 tx 50:     OK, {Fh 1 }
2021/12/28 01:48:28 rx 52: GETATTR n1 {Fh 0} 
2021/12/28 01:48:28 tx 52:     OK, {tA=0s {M040755 SZ=0 L=0 0:0 B0*4096 i0:0 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 54: READDIRPLUS n1 {Fh 1 [0 +4096)  L 0 DIRECTORY,NONBLOCK,0x8000} 
2021/12/28 01:48:28 tx 54:     OK,  160b data "\x02\x00\x00\x00\x00\x00\x00\x00"...
2021/12/28 01:48:28 rx 56: LOOKUP n1 ["file.txt"] 9b
2021/12/28 01:48:28 tx 56:     OK, {n2 g0 tE=0s tA=0s {M0100644 SZ=8 L=0 0:0 B8*4096 i0:2 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 58: GETATTR n2 {Fh 0} 
2021/12/28 01:48:28 tx 58:     OK, {tA=0s {M0100644 SZ=8 L=0 0:0 B8*4096 i0:2 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 60: LOOKUP n1 ["file.txt"] 9b
2021/12/28 01:48:28 tx 60:     OK, {n2 g0 tE=0s tA=0s {M0100644 SZ=8 L=0 0:0 B8*4096 i0:2 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 62: OPEN n2 {NONBLOCK,0x28000} 
2021/12/28 01:48:28 tx 62:     OK, {Fh 0 CACHE}
2021/12/28 01:48:28 rx 64: OPEN n2 {O_RDONLY} 
2021/12/28 01:48:28 tx 64:     OK, {Fh 0 CACHE}
2021/12/28 01:48:28 rx 66: IOCTL n2 *fuse._IoctlIn: &{{72 39 66 2 {{0 0} 837} 0} 0 0 2148034049 0 0 4} 
2021/12/28 01:48:28 tx 66:     38=function not implemented
2021/12/28 01:48:28 rx 68: RELEASE n2 {Fh 0   L0} 
2021/12/28 01:48:28 tx 68:     OK
2021/12/28 01:48:28 rx 70: FLUSH n2 {Fh 0} 
2021/12/28 01:48:28 tx 70:     OK
2021/12/28 01:48:28 rx 72: RELEASE n2 {Fh 0 NONBLOCK,0x28000  L0} 
2021/12/28 01:48:28 tx 72:     OK
2021/12/28 01:48:28 rx 74: READDIRPLUS n1 {Fh 1 [1 +4096)  L 0 NONBLOCK,DIRECTORY,0x8000} 
2021/12/28 01:48:28 tx 74:     OK
2021/12/28 01:48:28 rx 76: RELEASEDIR n1 {Fh 1 NONBLOCK,DIRECTORY,0x8000  L0} 
2021/12/28 01:48:28 tx 76:     OK
2021/12/28 01:48:28 rx 78: GETATTR n1 {Fh 0} 
2021/12/28 01:48:28 tx 78:     OK, {tA=0s {M040755 SZ=0 L=0 0:0 B0*4096 i0:0 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 80: STATFS n1 
2021/12/28 01:48:28 tx 80:     OK, {blocks (0,0)/0 files 0/0 bs0 nl0 frs0}
2021/12/28 01:48:28 rx 82: OPENDIR n1 
2021/12/28 01:48:28 tx 82:     OK, {Fh 1 }
2021/12/28 01:48:28 rx 84: GETATTR n1 {Fh 0} 
2021/12/28 01:48:28 tx 84:     OK, {tA=0s {M040755 SZ=0 L=0 0:0 B0*4096 i0:0 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 86: READDIRPLUS n1 {Fh 1 [0 +4096)  L 0 NONBLOCK,DIRECTORY,0x8000} 
2021/12/28 01:48:28 tx 86:     OK,  160b data "\x02\x00\x00\x00\x00\x00\x00\x00"...
2021/12/28 01:48:28 rx 88: LOOKUP n1 ["file.txt"] 9b
2021/12/28 01:48:28 tx 88:     OK, {n2 g0 tE=0s tA=0s {M0100644 SZ=8 L=0 0:0 B8*4096 i0:2 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 90: GETATTR n2 {Fh 0} 
2021/12/28 01:48:28 tx 90:     OK, {tA=0s {M0100644 SZ=8 L=0 0:0 B8*4096 i0:2 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 92: LOOKUP n1 ["file.txt"] 9b
2021/12/28 01:48:28 tx 92:     OK, {n2 g0 tE=0s tA=0s {M0100644 SZ=8 L=0 0:0 B8*4096 i0:2 A 0.000000 M 0.000000 C 0.000000}}
2021/12/28 01:48:28 rx 94: OPEN n2 {NONBLOCK,0x28000} 
2021/12/28 01:48:28 tx 94:     OK, {Fh 0 CACHE}
2021/12/28 01:48:28 rx 96: OPEN n2 {O_RDONLY} 
2021/12/28 01:48:28 tx 96:     OK, {Fh 0 CACHE}
2021/12/28 01:48:28 rx 98: IOCTL n2 *fuse._IoctlIn: &{{72 39 98 2 {{0 0} 838} 0} 0 0 2148034049 0 0 4} 
2021/12/28 01:48:28 tx 98:     38=function not implemented
2021/12/28 01:48:28 rx 100: RELEASE n2 {Fh 0   L0} 
2021/12/28 01:48:28 tx 100:     OK
2021/12/28 01:48:28 rx 102: FLUSH n2 {Fh 0} 
2021/12/28 01:48:28 tx 102:     OK
2021/12/28 01:48:28 rx 104: RELEASE n2 {Fh 0 NONBLOCK,0x28000  L0} 
2021/12/28 01:48:28 tx 104:     OK
2021/12/28 01:48:28 rx 106: READDIRPLUS n1 {Fh 1 [1 +4096)  L 0 DIRECTORY,NONBLOCK,0x8000} 
2021/12/28 01:48:28 tx 106:     OK
2021/12/28 01:48:28 rx 108: RELEASEDIR n1 {Fh 1 DIRECTORY,NONBLOCK,0x8000  L0} 
2021/12/28 01:48:28 tx 108:     OK
```